pkg/cloud/services/elb/loadbalancer.go

```diff
@@ -54,6 +55,340 @@ const maxELBsDescribeTagsRequest = 20
+		ClusterName: s.scope.Name(),
+		Lifecycle:   infrav1.ResourceLifecycleOwned,
+		Name:        aws.String(elbName),
+		Role:        aws.String(infrav1.APIServerRoleTagValue),
+		Additional:  s.scope.AdditionalTags(),
+	})
+
+	// If subnet IDs have been specified for this load balancer
+	if s.scope.ControlPlaneLoadBalancer() != nil && len(s.scope.ControlPlaneLoadBalancer().Subnets) > 0 {
+		// This set of subnets may not match the subnets specified on the Cluster, so we may not have already discovered them
+		// We need to call out to AWS to describe them just in case
+		input := &ec2.DescribeSubnetsInput{
+			SubnetIds: aws.StringSlice(s.scope.ControlPlaneLoadBalancer().Subnets),
+		}
+		out, err := s.EC2Client.DescribeSubnets(input)
+		if err != nil {
+			return nil, err
+		}
+		for _, sn := range out.Subnets {
+			res.AvailabilityZones = append(res.AvailabilityZones, *sn.AvailabilityZone)
+			res.SubnetIDs = append(res.SubnetIDs, *sn.SubnetId)
+		}
+	} else {
+		// The load balancer APIs require us to only attach one subnet for each AZ.
+		subnets := s.scope.Subnets().FilterPrivate()
+
+		if s.scope.ControlPlaneLoadBalancerScheme() == infrav1.ElbSchemeInternetFacing {
+			subnets = s.scope.Subnets().FilterPublic()
+		}
+
+	subnetLoop:
+		for _, sn := range subnets {
+			for _, az := range res.AvailabilityZones {
+				if sn.AvailabilityZone == az {
+					// If we already attached another subnet in the same AZ, there is no need to
+					// add this subnet to the list of the ELB's subnets.
+					continue subnetLoop
+				}
+			}
+			res.AvailabilityZones = append(res.AvailabilityZones, sn.AvailabilityZone)
+			res.SubnetIDs = append(res.SubnetIDs, sn.ID)
+		}
+	}
+
+	return res, nil
+}
+
+func (s *Service) createLB(spec *infrav1.LoadBalancer) (*infrav1.LoadBalancer, error) {
+	var t *string
+	switch s.scope.ControlPlaneLoadBalancer().LoadBalancerType {
+	case infrav1.LoadBalancerTypeNLB:
+		t = aws.String(elbv2.LoadBalancerTypeEnumNetwork)
+	case infrav1.LoadBalancerTypeALB:
+		t = aws.String(elbv2.LoadBalancerTypeEnumApplication)
+	case infrav1.LoadBalancerTypeELB:
+		t = aws.String(elbv2.LoadBalancerTypeEnumGateway)
+	}
+	input := &elbv2.CreateLoadBalancerInput{
+		Name:    aws.String(spec.Name),
+		Subnets: aws.StringSlice(spec.SubnetIDs),
+		Tags:    converters.MapToV2Tags(spec.Tags),
+		Scheme:  aws.String(string(spec.Scheme)),
+		Type:    t,
+	}
+	if s.scope.ControlPlaneLoadBalancer().LoadBalancerType != infrav1.LoadBalancerTypeNLB {
+		input.SecurityGroups = aws.StringSlice(spec.SecurityGroupIDs)
+	}
+
+	if s.scope.VPC().IsIPv6Enabled() {
+		input.IpAddressType = aws.String("dualstack")
+	}
+
+	out, err := s.ELBV2Client.CreateLoadBalancer(input)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create load balancer: %v", spec)
+	}
+
+	if len(out.LoadBalancers) == 0 {
+		return nil, errors.New("no new network load balancer was created; the returned list is empty")
+	}
+
+	// TODO: Add options to set up SSL.
```

Comment:
Is there a follow-up issue created for this todo?

Reply:
Ah, thanks for the reminder. There isn't yet. I will add that and paste the link there. TLS setup might be a bit disruptive to the code base and further complicate it. I'll do it in a follow-up PR.